src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
+            partition = (partition,)
+
+        if partition is not None:
+            if len(partition) > 2:
+                raise_error(
+                    ValueError,
+                    "A quantum channel should only contain one input system and one output system. "
+                    + "For general quantum networks, one should use the ``QuantumNetwork`` class.",
+                )
+            if len(partition) == 1:
+                if system_input == None:  # Assume the input is a quantum state
+                    partition = (1, partition[0])
+                else:
+                    if isinstance(system_input, bool):
+                        system_input = (system_input,)
+
+                    if system_input[0]:
+                        partition = (partition[0], 1)
+                    else:
+                        partition = (1, partition[0])
+
+        super().__init__(tensor, partition, pure=pure, backend=backend)
+
+    def is_unital(
+        self, order: Optional[Union[int, str]] = None, precision_tol: float = 1e-8
+    ):
+        """Returns bool indicating if the Choi operator :math:`\\mathcal{J}` of the network is unital.
+        A map is unital if it preserves the identity operator.
+
+        Unitality is calculated as distance between the partial trace of :math:`\\mathcal{J}`
+        and the Identity operator :math:`I`, with respect to a given norm.
+        Default is the ``Hilbert-Schmidt`` norm (also known as ``Frobenius`` norm).
+
+        For specifications on the other  possible values of the
+        parameter ``order`` for the ``tensorflow`` backend, please refer to
+        `tensorflow.norm <https://www.tensorflow.org/api_docs/python/tf/norm>`_.
+        For all other backends, please refer to
+        `numpy.linalg.norm <https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html>`_.
+
+        Args:
+            order (str or int, optional): order of the norm. Defaults to ``None``.
+            precision_tol (float, optional): threshold :math:`\\epsilon` that defines
+                if Choi operator of the network is :math:`\\epsilon`-close to unitality
+                in the norm given by ``order``. Defaults to :math:`10^{-8}`.
+
+        Returns:
+            bool: Unitality condition.
+        """
+        if precision_tol < 0.0:
+            raise_error(
+                ValueError,
+                f"``precision_tol`` must be non-negative float, but it is {precision_tol}",
+            )
+
+        if order is None and self._backend.__class__.__name__ == "TensorflowBackend":
+            order = "euclidean"
+
+        backend = self._backend
+
+        dim_out = self.partition[1]
+        dim_in = self.partition[0]
+
+        trace_out = backend.cast(
+            trace(dim_out, backend=backend).full(), dtype=self._tensor.dtype
+        )
+        trace_in = backend.cast(
+            trace(dim_in, backend=backend).full(), dtype=self._tensor.dtype
+        )
```

Comment:
Same thing here about casting

Reply:
Sure. I'll delete that

Reply:
I've updated the code regarding most of the comments. But I found this `cast` is still necessary as the data type between the trace operator and the target operator may not be the same (the trace operator only contains real number, but the target tensor can be complex)

Reply:
Even if the tensor contains only real numbers, its best to leave it as complex `dtype` by default in order to avoid these issues. If an user wants to change the `dtype` for whatever reason, leave it to them.